src/util/file_watch.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (C) 2022, Red Hat Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <string.h>
+#include <sys/time.h>
+#include <errno.h>
+#ifdef HAVE_SYS_INOTIFY_H
+#include <sys/inotify.h>
+#endif
+
+#include "util/inotify.h"
+#include "util/util.h"
+#include "util/file_watch.h"
+
+
+
+#define MISSING_FILE_POLL_TIME   10  /* seconds */
+#define FILE_WATCH_POLL_INTERVAL  5  /* seconds */
+
+
+struct file_watch_callback {
+    int wd;
+    time_t modified;
+    struct file_watch_callback *next;
+    struct file_watch_callback *prev;
+};
+
+struct file_watch_ctx {
+    struct tevent_context *ev;
+    const char *filename;
+    bool use_inotify;
+
+    struct config_file_inotify_check {
+        struct snotify_ctx *snctx;
+    } inotify_check;
+
+    struct config_file_poll_check {
+        TALLOC_CTX *parent_ctx;
+        struct tevent_timer *timer;
+        struct file_watch_callback *callbacks;
+    } poll_check;
+
+    fw_fn fn;
+    void *fn_arg;
+};
+
+
+static void poll_watched_file(struct tevent_context *ev,
+                             struct tevent_timer *te,
+                             struct timeval t, void *ptr);
+
+static errno_t create_poll_timer(struct file_watch_ctx *fw_ctx)
+{
+    struct timeval tv;
+
+    gettimeofday(&tv, NULL);
+    tv.tv_sec += FILE_WATCH_POLL_INTERVAL;
+    tv.tv_usec = 0;
+
+    fw_ctx->poll_check.timer = tevent_add_timer(fw_ctx->ev,
+                                                fw_ctx->poll_check.parent_ctx,
+                                                tv,
+                                                poll_watched_file,
+                                                fw_ctx);
+    if (!fw_ctx->poll_check.timer) {
+        talloc_free(fw_ctx);
```

Comment:
Feels like a weird place to free `fw_ctx`. Imo, it can result in a double free in `fw_watch_file::done` section.

Reply:
I fully agree. It's not right to free the context here. Let the caller do it if they want to.